src/bytes/bytes_lemmas.gobra

```diff
@@ -0,0 +1,359 @@
+	case b < a:
+		assert bytes.Repeat(s, a) == s ++ bytes.Repeat(s, a - 1)
+		lemmaSlicingRepeat(s, a-1, b-1)
+		assert bytes.Repeat(s, a-1)[: (b-1) * len(s)] == bytes.Repeat(s, MinInt(a-1, b-1))
+		assert bytes.Repeat(s, a)[: b * len(s)] == (s ++ bytes.Repeat(s, a - 1))[: b * len(s)]
+		assert 1 <= b
+		LemmaMulInjLeq(1, b, len(s))
+		assert len(s) <= b * len(s)
+		lemmaSliceSeqConcat(s, bytes.Repeat(s, a-1), b * len(s))
+		assert b * len(s) - len(s) == (b-1) * len(s)
+		assert (s ++ bytes.Repeat(s, a - 1))[: b * len(s)] == s ++ bytes.Repeat(s, a - 1)[:(b-1)*len(s)]
+		assert bytes.Repeat(s, a)[: b * len(s)] == bytes.Repeat(s, MinInt(a, b))
+	default:
+		assert false
+	}
+}
+
+
+// index
+
+ghost
+requires 2 <= len(s)
+requires t == len(s) - len(sep) + 1
+requires 0 <= i && i+1 <= t && t <= len(s)
+requires i < t
+requires s[i] != sep[0]
+requires forall j int :: { s[j:j+len(sep)] } 0 <= j && j < i ==> s[j:j+len(sep)] != sep
+requires forall j int :: { s[i+1:t][j] } 0 <= j && j < len(s[i+1:t]) ==> s[i+1:t][j] != sep[0]
+ensures  forall j int :: { s[j:j+len(sep)] } 0 <= j && j < t ==> s[j:j+len(sep)] != sep
+decreases
+func lemmaIndexIndexByteNotFound(s, sep seq[byte], i, t int) {
+	invariant i+1 <= idx
+	invariant forall j int :: { s[j:j+len(sep)] } 0 <= j && j < idx ==> s[j:j+len(sep)] != sep
+	decreases t-idx
+	for idx := i+1; idx < t; idx++ {
+		assert s[idx] == s[i+1:t][idx - (i+1)]
+		assert s[idx] != sep[0]
+		assert s[idx:idx+len(sep)] != sep
+	}
+}
+
+ghost
+requires forall j int :: { s[j:j+len(sep)] } 0 <= j && j < a ==> s[j:j+len(sep)] != sep
+requires forall j int :: { s[j:j+len(sep)] } a <= j && j < b ==> s[j:j+len(sep)] != sep
+ensures  forall j int :: { s[j:j+len(sep)] } 0 <= j && j < b ==> s[j:j+len(sep)] != sep
+decreases
+func lemmaIndexAdvance(s, sep seq[byte], a, b int) {
+
+}
+
+
+
+ghost
+requires InRangeInc(i, 0, len(s))
+requires InRangeInc(r, 0, len(s)-i)
+requires forall j int :: {s[i:][j:j+len(sep)]}{s[j]} 0 <= j && j < r ==> s[i:][j:j+len(sep)] != sep
+ensures  forall j int :: {s[j:j+len(sep)]} i <= j && j < r+i ==> s[j:j+len(sep)] != sep
+decreases
+func lemmaSliceSequence(s, sep seq[byte], i, r int) {
+	invariant forall j int :: {s[j:j+len(sep)]} i <= j && j < k ==> s[j:j+len(sep)] != sep
+	invariant InRangeInc(k, i, r+i)
+	decreases r+i-k
+	for k := i; k < r+i; k++ {
+		assert s[k] == s[i:][k-i]
+		assert s[k:k+len(sep)] === s[i:][k-i:k-i+len(sep)]
+		assert s[k:k+len(sep)] != sep
+	}
+}
+
+
+ghost
+requires InRangeInc(i, 0, len(s))
+requires InRangeInc(r, 0, len(s)-i)
+requires acc(sl.Bytes(s, 0, len(s)), R50)
+requires acc(sl.Bytes(s[i:], 0, len(s[i:])), R50)
+requires forall j int :: {sl.View(s[i:])[j:j+len(sep)]} 0 <= j && j < r ==> sl.View(s[i:])[j:j+len(sep)] != sep
+ensures  acc(sl.Bytes(s, 0, len(s)), R50)
+ensures  acc(sl.Bytes(s[i:], 0, len(s[i:])), R50)
+ensures  forall j int :: {sl.View(s)[i:][j:j+len(sep)]} 0 <= j && j < r ==> sl.View(s)[i:][j:j+len(sep)] != sep
+decreases
```

Comment:
Am I missing sth here, or your postconditions match some of your preconditions?

Reply:
The precondition operates on `sl.View(s[i:])`, whereas the postcondition on `sl.View(s)[i:]`. The `Bytes` pre/post conditions are repeated because they cannot be `preserves`, since a precondition depends on them and my LSP enforces a "pre < preserves < post" order.

That being said, a better lemma here would be `lemmaViewSubslice`, which asserts that `sl.View(s)[low:high] == sl.View(s[low:high])`. That lemma did not yet exist until very recently.

We could consider moving that lemma to `gobra-libs`

Reply:
So, I guess we can replace this implementation by a call with `lemmaViewSubslice`. Also, this lemma should be public